File: generic_sort.c
/* Name: Kent Tran
CECS 342 Assignment 5
Algorithm: generic sorting Language: C
Compiler: gcc
Editor: VSCode
Resources: https://www.geeksforgeeks.org/dsa/generic-implementation-of-quicksort-algorithm-in-c/
*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef struct
{
    const char *name;
    int age;
} Person;

int compareNums(const void *a, const void *b)
{
    double x = *(const double*)a;
    double y = *(const double*)b;
    if (x < y) return -1; 
    if (x > y) return 1;
    return 0;
}

int compareNames(const void *a, const void *b)
{
    Person *p1 = (const Person *)a;
    Person *p2 = (const Person *)b;
    return strcmp(p1->name, p2->name);
}

int compareAge(const void *a, const void *b)
{
    Person *p1 = (const Person *)a;
    Person *p2 = (const Person *)b;

    if (p1->age != p2->age)
        return p2->age - p1->age;

    return strcmp(p1->name, p2->name); // tie
}

void printNums(const double numbers[], int count)
{
    int i;
    for (i=0; i < count; i++)
    {
        printf("%.2f\n", numbers[i]);
    }
}

void printPeople(const Person people[], int count)
{
    int i;
    for (i = 0; i < count; i++)
    {
        printf("%s, %d\n", people[i].name, people[i].age);
    }
}

int main(void)
{
    double numbers[] =
    {
        645.41, 37.59, 76.41, 5.31, -34.23, 1.11, 1.10, 23.46, 635.47, -876.32, 467.83, 62.25
    };
    // gets total byte / 4(byte of first array only)
    int num_count = sizeof numbers / sizeof numbers[0];

    Person people[] = 
    {
        {"Hal", 20}, {"Susann", 31}, {"Dwight", 19}, 
        {"Kassandra", 21}, {"Lawrence", 25}, {"Cindy", 22}, 
        {"Cory", 27}, {"Mac", 19}, {"Romana", 27}, {"Doretha", 32}, 
        {"Danna", 20}, {"Zara", 23}, {"Rosalyn", 26},
        {"Risa", 24}, {"Benny", 28}, {"Juan", 33}, {"Natalie", 25}
    };
    int people_count = sizeof(people) / sizeof(people[0]);

    Person peopleByName[17];
    Person peopleByAge[17];

    memcpy(peopleByName, people, sizeof(people));
    memcpy(peopleByAge, people, sizeof(people));

    qsort(numbers, num_count, sizeof(double), compareNums);
    qsort(peopleByName, people_count, sizeof(Person), compareNames);
    qsort(peopleByAge, people_count, sizeof(Person), compareAge);

    printf("Sorted numbers (ascending):\n");
    printNums(numbers, num_count);

    printf("\nSorted people by name:\n");
    printPeople(peopleByName, people_count);

    printf("\nSorted people by age (descending, then name):\n");
    printPeople(peopleByAge, people_count);

    return 0;
}

File: generic_sort.cs
/* Name: Kent Tran
CECS 342 Assignment 6
Algorithm: generic sorting Language: C#
Compiler:
Editor: VSCode
Resources: 
*/

using System;
using Systen.Collections.Generic;
using System.Linq;

class Person
{
    public string Name { get; set; }
    public int Age { get; set; }

    public Person(string name, int age)
    {
        Name = name;
        Age = age;
    }
}

class Program
{
    static void Main()
    {
        List<double> numbers = new List<double>
        {
            645.41, 37.59, 76.41, 5.31, -34.23, 1.11, 1.10, 23.46, 635.47, -876.32, 467.83, 62.25
        };

        List<Person> people = new List<double>
        {
        new Person("Hal", 20), new Person("Susann", 31), new Person("Dwight", 19), 
        new Person("Kassandra", 21), new Person("Lawrence", 25), new Person("Cindy", 22), 
        new Person("Cory", 27), new Person("Mac", 19), new Person("Romana", 27), new Person("Doretha", 32), 
        new Person("Danna", 20), new Person("Zara", 23), new Person("Rosalyn", 26),
        new Person("Risa", 24), new Person("Benny", 28), new Person("Juan", 33), new Person("Natalie", 25)
        };

        List<double> sortedNumbers = numbers.OrderBy(n => n).ToList();

        List<Person> sortedNames = people.OrderBy(p => p.name).ToList();

        List<Person> sortedAges = people
            .OrderByDescending(p => p.age)
            .ThenBy(p => p.Name)
            .ToList();

        Console.WriteLine("Sorted numbers: ");
        foreach (double n in sortedNumbers)
            Console.WriteLine($"{n:F2}");

        Console.WriteLine("\nSorted by name: ");
        foreach (Person p in sortedNames)
            Console.WriteLine($"{p.Name}, {p.Age}");

        Console.WriteLine("\nSorted by age, then name: ")
        foreach (Person p in sortedAges)
            Console.WriteLine($"{p.Name}, {p.Age}");
    }
}